initial re-formatting; add .clang-format files

Mmessiter opened the pull request:
resolves #4

2bndy5 commented:
This is still in the testing phase so I marked the PR as a draft to avoid accidentally merging it.

2bndy5 commented:
You shouldn't need to put all the libraries back. In fact if all you need is to roll back the RF24 lib, then that can be done by changing 1 line in the RXcode's platformio.ini file under the `lib_deps` option.

https://github.com/Mmessiter/LockDownRadioControl/blob/4c36cbddfd63ae1173dd7a3414e8d9aad2bd04dc/ReceiverCode/platformio.ini#L25

which can be changed to
```ini
    nrf24/RF24 @ 1.1.5
```

Mmessiter commented:
Yes - sorry! I guessed it would be something along those lines. But I didn't quite know so I rather over did it just to get it working. By the way, where in the RF24 library is the version number? I didn't find it yet.

2bndy5 commented on ReceiverCode/src/main.cpp:
```diff
@@ -1,1105 +1,1209 @@
+    InitCurrentRadio();
+#endif
+#ifndef SECOND_TRANSCEIVER
+    Radio2 = Radio1;
```
You shouldn't need to do this. In fact, this may cause problems when you switch `CurrentRadio` from `Radio1` to `Radio2` (which would just be a duplicate of `Radio1` after this line).

Mmessiter replied:
Ah thank you! 
I learn a little every day!

2bndy5 commented on ReceiverCode/src/main.cpp:
```diff
@@ -1,1105 +1,1209 @@
+        CurrentRadio->stopListening();
+        if (Rnumber == 1) {
+            Rnumber      = 2;
+            CurrentRadio= &Radio2;
```
inconsistent syntax. missing a space before the `=`.

Mmessiter commented:
You're right. Currently only Radio1 is used. 
I need to make up boards with two ML01DP5 transceivers. These are much better than ML01SP4 which were simply not good enough. When I have these I'll get both working.

Mmessiter commented:
I'll fix that 😁!

2bndy5 commented on ReceiverCode/src/main.cpp:
```diff
@@ -1,1105 +1,1209 @@
+    ScanI2c(); // see what's connected
+    if (USE_BNO055) BNO055_Cheapo.begin();
+    if (USE_BNO055A) BNO055_Adafruit.begin();
+    SPI.begin();
+    SPI.setClockDivider(SPI_CLOCK_DIV16); // for Teensy - Slow SPI bus
```
Are there any other modules using the SPI bus?

Calling `RF24::begin()` invokes `SPI.begin()`. According to the Teensy SPI docs, using `SPI.setClockDivider()` is deprecated.
>SPI speed was set indirectly, as a function of the Teensy clock, with SPI.setClockDivider(divider). SPI_CLOCK_DIV2 was the fastest option. This meant that code running on a 16MHz Teensy 2 and a 96MHz Teensy 3.2 would set the SPI speed differently to achieve same actual speed for the device.

All this means is if you want to slow the SPI bus speed, then you could simply specify that to the RF24 c'tor
```cpp
RF24 Radio1(pinCE1, pinCSN1, 6000000); // uses a SPI speed of 6MHz (default is 10MHz)
```
but that makes the SPI speed of 6MHZ only used by the instantiated RF24 object. If there are other SPI devices on that bus, then it is likely changing the SPI speed to the default specified by the said device's driver lib. RF24 lib uses SPI lib's `beginTransaction()` to ensure the SPI bus' speed is consistent when using the radio.

So, do you need to change the SPI bus' clock divider?

Mmessiter replied:
No other devices are using the SPI bus. 

I had not given this a thought for some time. At first I could not get it to work on the Teensy 4.x - until I discovered that the cause was excessive bus speed - so I slowed it down. It seems to work well now. I dare say there are indeed other ways to get it to work - but do you think it should be implemented in a different way?

2bndy5 replied:
I already stated the currently supported way to implement changes to the SPI bus speed; I would recommend you remove these 2 lines and adjust the RF24 c'tor calls.

Generally, if the default 10MHz SPI speed doesn't work correctly, then this means that the SPI bus connections between the radio module and the MCU are (1 or more of the following conditions)
- too long (causing "voltage drop" or "cross-talk")
- too low of a gauge (mainly when using jumper wires but may also be caused by custom PCB rountings' width)
- poorly terminated (using jumper wires or inadequate solder joints on the custom PCB)

Mmessiter replied:
I ran some tests. Yes the explicit SPI.begin(); and the setting of the clock divisor both appear now to be quite unnecessary as you said. I have commented these lines out. It changed nothing. 

Then I put the new library back in, and hoped, and tested. 

Sadly it was still just as before - nearly working but with lots of lost packets exactly as before. So I guess it's a something else. I feel we were very close to the answer! 

I added that missing space too.

2bndy5 replied:
ok I'll keep digging

2bndy5 commented on ReceiverCode/src/main.cpp:
```diff
@@ -1,1105 +1,1209 @@
+{
+    pinMode(LED_PIN, OUTPUT);
+    digitalWrite(LED_PIN, HIGH);
+    delay(2000); // Needed ! - possibly for stabilising capacitors.
+    Serial.begin(9600);
```
For MCUs with native USB support (like the Teesny), I recommend adding a 
```cpp
while (!Serial) {}
```
after `Serial.begin();` call in `setup()`. 

However, I'm uncertain what problem the `delay(2000)` is aiming to solve. Usually (in my experience with ATSAMD51 boards), without this `while (!Serial)` loop, much of the initial Serial output is not actually sent over USB.

Mmessiter commented:
Thanks - I forget now why exactly I added in that two second pause. I will try as you suggest and test. 

I also just removed those same two unnecessary SPI lines from TransmitterCode.

Mmessiter commented:
I admire your tenacity! It's powerful stuff when combined with experience and great knowledge! 
I have the tenacity - but the other two are still a work in progress!

Mmessiter commented:
Did you also observe that an overfed dog will explode ? 😂

2bndy5 commented:
Technically, you cannot disable CRC if using ACK packets. The radio's firmware will force CRC on when auto-ack is enabled. I only mention this because I keep seeing the comment `could be 16 or disabled`

2bndy5 commented on TransmitterCode/src/main.cpp:
```diff
@@ -1,4999 +1,5836 @@
-                   if (Channel12Switch == SwitchEditNumber) Channel12Switch=0;}
-        
-                if  (SwitchEditNumber==1)  {if (GetValue(OneSwitchViewc_revd)) {Switch1Reversed=true;} else {Switch1Reversed=false;}}
-                if  (SwitchEditNumber==2)  {if (GetValue(OneSwitchViewc_revd)) {Switch2Reversed=true;} else
```
`isAckPayloadAvailable()` is deprecated. Use `available()` instead (these 2 functions are synonomous under the hood).

2bndy5 commented:
You seem to call `InitRadio()` a lot. I don't think this is a problem, but calling `RF24::begin()` and then customizing the radio's config is quite time-consuming (uneeded SPI transactions slow things down considerably).

Mmessiter commented:
Thanks - yes of course

Mmessiter commented:
Ok I'll change that ...

Mmessiter commented:
That's because I was working getting two transceivers to take over from one another. When that bit is done I hope it'll be more streamlined. But I must get new PCBs ...

Mmessiter commented:
Good idea to take a break! We will crack it eventually and I bet it's something simple we've just not seen yet. By all means remove libraries as before - provided the working version of the RF24 one is still in. 

I find it's often when taking a break and a long walk that my subconscious mind often comes up with the answer that I just could not see while consciously searching.

2bndy5 commented:
I reverted the commit that added all the libs and rolled back the version of RF24 lib in RXcode (via platformio.ini - I removed copy of RF24 lib). Turns out that v1.1.5 of RF24 is not available in PIO registry. The earliest version of RF24 available for PIO is v1.1.7, so I bumped it up to that.

2bndy5 commented:
If you find that v1.1.7 of RF24 lib still has problems than that would narrow down the possible culprits. However, I just looked through a [comparison of the source code for RF24 from v1.1.5 to v1.1.7](https://github.com/nRF24/RF24/compare/v1.1.5...v1.1.7#files_bucket), and its mostly changes to linux build system (& some renaming of internal/private variables).

2bndy5 commented:
BTW, there's a reason I have duplicate _.clang-format_ files (needs to be 1 for each _platformio.ini_ file)

Mmessiter commented:
That's great thank you. I'll check it....

Mmessiter commented:
There's still quite a few things I don't fully understand about this complex environment so I hope I don't blunder too frequently! I'm learning - but it may appear agonisingly slow.

2bndy5 commented:
To be honest, I'm only familiar with using PIO to develop libraries. I've never had a reason to use PIO for a personal project. Although it makes more sense to use PIO for the teensy because there's less setup for them in PIO than what's required to setup teensyduino for the Arduino IDE. In fact, I really only use PIO to test compile the RF24 libs for the teensy boards (cannot setup teensyduino on CLI for an Arduino IDE based CI workflow -- 😨 attack of the acronyms).

Mmessiter commented:
I like PIO a lot - I moved to it about 6 months ago. To be fair, it has corners and features which I just leave on default because I've had no need to change them - and don't really understand them either.

2bndy5 commented:
any problems with RF24 v1.1.7?

Mmessiter commented:
Late breakfast still in progress here sorry! I'll check immediately afterwards and report immediately.

2bndy5 commented:
no worries 
# 🧇

Mmessiter commented:
Ok! Breakfast paused - check run - no problem with that version - works perfectly!😄

2bndy5 commented:
resume breakfast before it gets cold!

2bndy5 commented:
You just pushed a 4 copies of RF24 lib back into this branch. Please make sure your local copy of the branch is up-to-date with the remote copy before making a commit. This is a common mistake when more than 1 person is working on the same branch. I can fixed this, but now I'm worried that you didn't actually test v1.1.7 because you didn't update your local copy of the branch.

Mmessiter commented:
Pardon me. I'll check that again ...

2bndy5 commented:
Its all good. I remember it took me quite a while (almost a year of off-n-on usage) to get used to using github.

Mmessiter commented:
Thank you for your tolerance! I have now removed my surplus copies of the RF24 library and tested again. Happily all is still working perfectly!

I really want to work on the documentation for building the radio today. I have downloaded Macdown￼ which appears by reputation to be a good MD editor. But I'm struggling to use it especially with positioning images. Can you recommend a good editor and/or method?￼

2bndy5 commented:
> Can you recommend a good editor and/or method?￼

Writing documentation is not a user friendly thing for me. I worked a year in college as a english tutor, so I'm not as bad at writing anything than others. In general, I write all docs from vscode's editor and use some extensions to check/render the finished product before I commit. After a while, I just found myself struggling with spelling errors.

I recently upped my docs game by using graphviz software to generate graphs ([see this page as my triumph](https://circuitpython-nrf24l01.readthedocs.io/en/rf24-network/network_docs/topology.html)), but that's an advanced skillset.

Mmessiter commented:
... Becoming familiar with new syntax and GitHub Probably will take me a little more time than most because among other things I'm a bit dyslexic! ￼(I read music much more fluently than words.)

New uses for old words (eg. "Depreciated") need (for me) to be unambiguously defined before I can properly grasp the intended meaning. Perhaps there's an online Git jargon translator out there somewhere?! 😁

Mmessiter commented:
… That page is a triumph indeed! Bravissimo!￼

2bndy5 commented:
>I'm a bit dyslexic

yikes! my mom didn't know she was dyslexic for the first 15 years of her life. I can't even imagine such a hardship.

ps - I haven't mentioned this yet, but I took 2 years of music theory in High school (20 years ago), and have since stopped reading sheet music. I can play guitar, bass and piano (if I practice enough - I'm very rusty). I also apprenticed for a self-employed electric violin maker a couple years ago; that was really cool because we did everything by hand and outsourced custom designed inlays (fingerboards included though inlayed fingerboards are usually made of carbon fiber vs the usual ebony).. As a rookie, I got stuck doing a lot of pro-bono work for local high school's band instrument (mostly acoustic basses and cellos). Hint: strong superglue and wood dust are better than wood putty.

2bndy5 commented:
>  But I'm struggling to use it especially with positioning images.

Not that I would recommend it, but md syntax allow for raw html usage as well. Although rendering md with html is largely dependent on the software used to view it.

For example (from the project I've been working on today):
~~~
ok I've decided on this:
<!-- cpp linter action -->
## :scroll: Run `clang-format` on the following files
- [ ] demo/demo.cpp
- [ ] demo/demo.hpp

---
## :speech_balloon: Output from `clang-tidy`
<details><summary>demo/demo.cpp</summary><br>
<details open>
<summary><strong>demo/demo.cpp:3:10:</strong> warning: [modernize-deprecated-header]

> inclusion of deprecated C++ header 'stdio.h'; consider using 'cstdio' instead
</summary><p>

```cpp
#include <stdio.h>
         ^~~~~~~~~
         <cstdio>
```
</p>
</details>

<details open>
<summary><strong>demo/demo.cpp:8:5:</strong> warning: [modernize-use-trailing-return-typ]

> use a trailing return type for this function
</summary><p>

```cpp
int main(){
~~~ ^
auto       -> int
```
</p>
</details>

<details open>
<summary><strong>demo/demo.cpp:10:13:</strong> warning: [readability-braces-around-statement]

> statement should be inside braces
</summary><p>

```cpp
    for (;;) break;
            ^
             {
```
</p>
</details>

<details open>
<summary><strong>demo/demo.cpp:13:5:</strong> warning: [cppcoreguidelines-pro-type-varar]

> do not call c-style vararg functions
</summary><p>

```cpp
    printf("Hello world!\n");
    ^
```
</p>
</details>

<hr></details><details><summary>demo/demo.hpp</summary><br>
<details open>
<summary><strong>demo/demo.hpp:6:11:</strong> warning: [modernize-use-default-member-ini]

> use default member initializer for 'useless'
</summary><p>

```hpp
    char* useless;
          ^
                 {"\0"}
```
</p>
</details>

<details open>
<summary><strong>demo/demo.hpp:7:9:</strong> warning: [modernize-use-default-member-ini]

> use default member initializer for 'numb'
</summary><p>

```hpp
    int numb;
        ^
            {0}
```
</p>
</details>

<details open>
<summary><strong>demo/demo.hpp:11:11:</strong> warning: [modernize-use-trailing-return-typ]

> use a trailing return type for this function
</summary><p>

```hpp
    void *not_usefull(char *str){useless = str;}
    ~~~~~~^
    auto                         -> void *
```
</p>
</details>

</details>
~~~
randers like this
ok I've decided on this:
<!-- cpp linter action -->
## :scroll: Run `clang-format` on the following files
- [ ] demo/demo.cpp
- [ ] demo/demo.hpp

---
## :speech_balloon: Output from `clang-tidy`
<details><summary>demo/demo.cpp</summary><br>
<details open>
<summary><strong>demo/demo.cpp:3:10:</strong> warning: [modernize-deprecated-header]

> inclusion of deprecated C++ header 'stdio.h'; consider using 'cstdio' instead
</summary><p>

```cpp
#include <stdio.h>
         ^~~~~~~~~
         <cstdio>
```
</p>
</details>

<details open>
<summary><strong>demo/demo.cpp:8:5:</strong> warning: [modernize-use-trailing-return-typ]

> use a trailing return type for this function
</summary><p>

```cpp
int main(){
~~~ ^
auto       -> int
```
</p>
</details>

<details open>
<summary><strong>demo/demo.cpp:10:13:</strong> warning: [readability-braces-around-statement]

> statement should be inside braces
</summary><p>

```cpp
    for (;;) break;
            ^
             {
```
</p>
</details>

<details open>
<summary><strong>demo/demo.cpp:13:5:</strong> warning: [cppcoreguidelines-pro-type-varar]

> do not call c-style vararg functions
</summary><p>

```cpp
    printf("Hello world!\n");
    ^
```
</p>
</details>

<hr></details><details><summary>demo/demo.hpp</summary><br>
<details open>
<summary><strong>demo/demo.hpp:6:11:</strong> warning: [modernize-use-default-member-ini]

> use default member initializer for 'useless'
</summary><p>

```hpp
    char* useless;
          ^
                 {"\0"}
```
</p>
</details>

<details open>
<summary><strong>demo/demo.hpp:7:9:</strong> warning: [modernize-use-default-member-ini]

> use default member initializer for 'numb'
</summary><p>

```hpp
    int numb;
        ^
            {0}
```
</p>
</details>

<details open>
<summary><strong>demo/demo.hpp:11:11:</strong> warning: [modernize-use-trailing-return-typ]

> use a trailing return type for this function
</summary><p>

```hpp
    void *not_usefull(char *str){useless = str;}
    ~~~~~~^
    auto                         -> void *
```
</p>
</details>

</details>

Mmessiter commented:
Gosh! You might enjoy this as light relief:

https://youtu.be/UbwfAc0AKhk

Ten years ago I built a machine to play a real harpsichord automatically using MIDI. It's my robot accompanist when I play oboe. This morning my wife and I played a trio! Two oboes one automatic harpsichord😁

2bndy5 commented:
> I experimented with different library versions - then to get it working I had to simply library use again because just reverting the ini file back to 1.1.7 didn't work.
I don't know why!

How exactly did you experiment with different library versions? What versions did you try (& what was successful)? Did you try deleting the _.pio_ folder before rebuilding v1.1.7?

If you want to revert to using the RF24 lib, you don't need to copy all the libs back to do that.

Mmessiter commented:
Good morning Brendan I hope you're well! 

... Yes I wish I understood more. 
I'm so sorry...

Here's what happened:

I saw the reference to 1.1.7 in ini file. 
I wanted to make a test with 1.4.1 so I simply edited that number to 1.4.1 and did the test. 

(My test was to change the timing a little but only when hopping to another channel...)

My edit appeared successfully to invoke the later library. But my test didn't fix the problem. So I reverted my code and edited that line again back to 1.1.7 assuming this would once again use the older library. But that didn't work as expected! The compiler gave **hundreds** of error messages. At this point I was left with a non-usable version, and the only way I knew would fix it was to have intended libraries physically there without references to hidden versions anywhere elsewhere. 

(I suggest that this is perhaps a simpler way to work when trying out different libraries. Clearly I don't yet understand just how the hidden ones are invoked. )

By the way, the 'main' branch is being ignored now. How do we make the current branch into 'main' ? 

Yesterday I also designed a new PCB which uses two ML01DP5 transceivers instead of ML01SP4. These I find are **far** better. But bigger. JLCPCB in China should get these to me within a week or so - then I hope to complete the two-transceivers version of Receive !

... and I learned a lot yesterday about MD format (new to me) and I now use VSCODE for this because of some great addons it has that simulate GitHub, allow pasting images, etc.!

2bndy5 commented:
> The compiler gave hundreds of error messages

Thank you for stepping me through your process. I was able to replicate the compiler errors, and they all relate to having 2 built copies of the RF24 lib in the _.pio/build/_ directory. Its very common to have to delete the contents of a project's build directory between consecutive build attempts (its practically required when using CMake). I would recommend deleting the _.pio_ folder when you first encounter build problems.

>By the way, the 'main' branch is being ignored now. How do we make the current branch into 'main' ?

The main purpose of a pull request to to **request** the repo maintainer(s) to **pull** changes from a branch. When you say "the current branch", you're referring to the branch that your local copy of the repo is using. If you want to merge this branch into the repo's main branch, then that's what this pull request is for...
![image](https://user-images.githubusercontent.com/14963867/134630380-d0dbdc3f-0393-4cdd-a4f6-6a120bb19e31.png)
Please mind the dark theme (its a screen shot of the top of this page). Notice the fine print in the picture.
Now notice the bottom of this page:
![image](https://user-images.githubusercontent.com/14963867/134630782-9b3f0b1b-407e-4116-9d7a-16aced43a059.png)
I've changed this PR's status to "draft" because you and I have been playing with experimental changes on this branch.

Mmessiter commented:
Thank you. I may learn eventually!

You say 'that is what the pull request is for' but I still don't know how physically to make a branch into the main one - nor indeed what would happen if someone had modified the main branch in the meantime. Are these edits lost? How do we keep them? So currently I dare not try! 

Also I've just searched for a .pio folder but didn't find one. Is it hidden? Where is it? 

Please pardon my profound ignorance!

Mmessiter commented:
Ok I just found the .pio folder! It was indeed hidden - and has a lot in it! 

So I guess in future I should edit the version number (as I did) AND then delete this entire hidden .pio folder before recompiling?

2bndy5 commented:
>Also I've just searched for a .pio folder but didn't find one. Is it hidden? Where is it?

Any folder that starts with a `.` is a hidden folder; this a convention of all OSs (I usually have my operating system set to not hide hidden folders/files). It is created when you execute `pio run`. For me it exists in the folder that contains the _platformio.ini_ file.

> I still don't know how physically to make a branch into the main one - nor indeed what would happen if someone had modified the main branch in the meantime. Are these edits lost? How do we keep them? So currently I dare not try!

These questions are best left to [github docs' quickstart](https://docs.github.com/en/get-started/quickstart/github-flow) because they're such a basic concepts of github (and all Version Control Services - "VCS" for short).

Mmessiter commented:
Thanks! I will read up....
And thank you for your patience!

2bndy5 commented:
It might be better to only call `RF24:: begin()` just once in `setup()` instead of everyone `InitRadio()` is called. `RF24:: begin()` does a lot more than just perform a soft reset of the radio's registers... It initializes the CE & CSN pins, calls `SPIClass::begin()`, and returns a bool describing if the radio responded correctly to the soft reset of the registers.

When moving `RF24::begin()` to only in `setup()`, the default settings in `InitRadio()` should be hard coded into the `switch...case` blocks that configure the radio's settings. It might be best set the `default` case to set the radio setting to it's default state. Defaults should be stated in the RF24 docs, but I can help if you need clarification.

Mmessiter commented:
I have today modified code so that InitCurrentRadio() is ONLY called from Setup() - *never* again. So I think this achieves what you recommend. Hope so. 

Among other reasons, I realise now that it takes quite a while to initialise - and this could cause a dangerous gap in data communication so it mustn't happen.

Incidentally I really appreciate your help. Thank you! 

I was incommunicado for several hours today - sorry. Reason was: first I flew my helicopter in beautiful autumn sunshine. And then a new iPhone 13 showed up so I had to set it up ! 😁

2bndy5 commented:
>I have today modified code so that InitCurrentRadio() is ONLY called from Setup() - never again. So I think this achieves what you recommend. Hope so.

I would agree.

>Among other reasons, I realise now that it takes quite a while to initialise - and this could cause a dangerous gap in data communication so it mustn't happen.

Same goes for the transmitter... You seem to use `InitRadio()` to set the TX pipe, so maybe that should be abstracted as well.

Mmessiter commented:
I'll check that! If transmitter calls init more than once I'll fix it! Thank you.

Mmessiter commented:
Ah yes - the 'binding' process must change the pipe. I thought that required a new init call - but maybe the pipe can be changed without a re-init...

2bndy5 commented:
Just make sure to call `RF24::stopListening()` **after** calling `RF24::openWritingPipe()`; this is a common gotcha for the RF24 API.

Mmessiter commented:
Thanks - will do!

2bndy5 commented:
BTW, I did some retro-releases for all the RF24 libs to the PIO lib registry. [As you can see here](https://platformio.org/lib/show/433/RF24/installation), v1.1.5 is now available for install via PIO. However, the latest release (on PIO its v1.3.12) isn't actually the newest release (on github its v1.4.1), but that will be corrected upon release of v1.4.2 (which has been pending for months now).

I hope that with this knowledge and the factoid about clearing the _.pio_ folder between consecutive build attempts will help avoid keeping copies of libs in this repo.

Mmessiter commented:
Thank you - yes now at least I know to delete that .pio path next time I see hundreds of compiler errors!

Mmessiter commented:
This morning I implemented a lot of improvements to the receiver code - in a different directory. I didn't want to screw up the one linked to GitHub! Then I simply copied it over the one which is linked to GitHub. And I tried to sync it with GitHub. Github complained of conflicts.￼ And gave me a number of messages which I didn't understand! So now I can't update that one on GitHub but it looks good in my house! So sorry!￼

2bndy5 commented:
Conflicts are very common. I still have to remedy those situations, but it is usually caused by many changes. Just try to copy the _src/main.cpp_. How are you synching your local copy of the repo with the remote copy on github? Are you using the "github for desktop" app? Or are you using the git bash CLI?

2bndy5 commented:
If I close this PR then it might be easier to copy the over the updates.

Mmessiter commented:
I'm using the desktop app on iMac ... I'll try again in a moment!

Mmessiter commented:
Bravo! Thank you! Now I have updated that code successfully. You're a magician 😁

.... now I'll attend to transmitter code ...

Mmessiter commented:
I have edited and tested both transmitter and receiver code, and successfully synched these with GitHub. ALL surplus calls to initRadio are gone now - and replaced with a simpler pipe change function. 
It works fine! Thank you! Still on older library though ...

2bndy5 commented:
>Still on older library though ... 

damn. Well at least the code should run a lot smoother.

Mmessiter commented:
Yes it seems to have attained an excellent new level of robustness. And I'm eager to try the new PCBs I designed the other day because I've thought of a better way to use two (or more) transceivers. Since both are pre-initialised, when scanning to reconnect after lost packets it can simply lock into whichever transceiver is first to see the transmitter. That might be very quick.

The ML01SP4s were tiny, but not quite so good. So I want to try it with multiple ML01DP5s.
